### DISCUSSION TITLE
Error when executing client for pt1 of challenge

### DISCUSSION THREAD

**Discussion by David-Feldt** [uwrobotics/uwrt_mars_rover#161] [Q&A]
Hi I get this error when trying to execute my client
```
terminate called after throwing an instance of 'rclcpp::exceptions::RCLInvalidArgument'
  what():  failed to create interrupt guard condition: context argument is null, at /tmp/binarydeb/ros-foxy-rcl-1.1.11/src/rcl/guard_condition.c:65

```
What could I be doing wrong?

**Comment by wmmc88** [comment]
Did you call `rclcpp::init` in your node?

**Reply by David-Feldt** [reply]
I did not

**Reply by wmmc88** [reply]
Ah. Try that. I think it should resolve the error. 

There may be other things missing in ur code. Have another look at the rclcpp node examples in the ros2 tutorials.

**Reply by David-Feldt** [reply]
I needed to add the argc and argv to my main and init, turtle has been killed

**Reply by niiquaye** [reply]
Please take a look at using composition.

**Comment by niiquaye** [comment]
Can you provide a code snippet?

**Reply by David-Feldt** [reply]
`#include "rclcpp/rclcpp.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <turtlesim/srv/kill.hpp>

using namespace std::chrono_literals;

int main()
{
  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("kill_client");
  rclcpp::Client<turtlesim::srv::Kill>::SharedPtr client = node->create_client<turtlesim::srv::Kill>("/kill");

  auto request = std::make_shared<turtlesim::srv::Kill::Request>();
  request->name = "turtle1";

  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "service not available, waiting again...");
  }

  auto result = client->async_send_request(request);
  if (rclcpp::spin_until_future_complete(node, result) ==
      rclcpp::FutureReturnCode::SUCCESS)
    {
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Turtle killed");
    } else {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Failed to kill");
    }

    rclcpp::shutdown();
   return 0;
}
`

**Reply by niiquaye** [reply]
You are using the old ROS1 way of instantiating a client. Take a look at how to create a component. Furthermore, look at the line when you are creating a client, you are missing an angle bracket for the template argument.

**Reply by niiquaye** [reply]
Take a look at [this](https://github.com/ros2/examples/tree/master/rclcpp/composition/minimal_composition).
